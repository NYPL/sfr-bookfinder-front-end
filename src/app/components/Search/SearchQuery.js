import axios from 'axios';
import appConfig from '../../../../appConfig';

function performSearch(query, filter) {
  // Need a filter to use; default to 'q'
  const defaultFilter = filter;
  // Need a parsed query input to use for each filter
  const userQuery = (query) ? encodeURI(query) : 'test';
  // Need a client to send the search and receive results
  // Need to pass the results to a renderer
  const apiUrl = appConfig.esUrl.development;
  let queryResp = {};

  queryResp = axios.get(apiUrl, { params: { q: userQuery } })
    .then((resp) => {
      if (resp.data) {
        // Update the search results object?
        const response = resp.data;
        return response;
      }
    })
    .catch((err) => {
      console.log('Query error', err);
    });

  return queryResp;
}

export default performSearch;
