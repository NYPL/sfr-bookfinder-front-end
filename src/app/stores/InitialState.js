import PropTypes from 'prop-types';

export const initialSearchQuery = {
  query: '',
  field: 'keyword',
  showQuery: '',
  showField: '',
  per_page: 10,
  page: 0,
  total: 0,
  filters: [],
};

export const searchQueryPropTypes = PropTypes.shape({
  query: PropTypes.string,
  field: PropTypes.string,
  showQuery: PropTypes.string,
  showField: PropTypes.string,
  per_page: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  page: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
});

const initialState = {
  searchResults: {},
  searchQuery: initialSearchQuery,
  sort: {
    sortFilter: 'title',
    sortOrder: 'asc',
  },
  workDetail: { instances: [] },
};

export default initialState;
