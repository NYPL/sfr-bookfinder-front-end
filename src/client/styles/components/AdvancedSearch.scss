.sub-legend {
  font-size: 1rem;
  text-transform: uppercase;
  font-weight: normal;
  border-bottom: 1px solid black;
  display: block;
  width: 100%;
  line-height: 2rem;
}

.nypl-select__menu {
  @media (min-width: 40em) {
    margin-right: 1rem;
  }
  @media (min-width: 64em) {
    margin-right: 2rem;
  }
}
.nypl-select {
  @media (min-width: 40em) {
    margin-right: 1rem;
  }
  @media (min-width: 64em) {
    margin-right: 2rem;
  }
}
.nypl-select__multi-value {
  border: 1px dashed #7fb4dc;
}

.tag-button {
  color: black;
  box-shadow: inset 0 0 0 1px #000;
  border-radius: 0;
  position: relative;
  padding-right: 1.5rem;
  cursor: pointer;

  &:after {
    display: block;
    position: absolute;
    top: 0;
    right: 0;
    font-size: 0.7rem;
    content: 'x';
    padding: 0.5rem 0.5rem;
  }
}
