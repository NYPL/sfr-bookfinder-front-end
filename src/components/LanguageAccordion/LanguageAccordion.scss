.language-accordion { 

    .accordion { 
        background-color: var(--ui-white);
    }

}
