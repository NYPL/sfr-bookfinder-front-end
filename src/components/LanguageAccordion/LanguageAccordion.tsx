import { FacetItem } from "~/src/types/DataModel";
import * as DS from "@nypl/design-system-react-components";
import React from "react";
import { Filter } from "~/src/types/SearchQuery";

// An Accordion of languages

const LanguageAccordion: React.FC<{
  languages: FacetItem[];
  showCount: boolean;
  selectedLanguages: Filter[];
  onLanguageChange: any;
}> = (props) => {
  const { languages, showCount, selectedLanguages, onLanguageChange } = props;

  const selectedLanguageFilter = (language: string) => {
    return selectedLanguages.find((langFilter) => {
      return langFilter.value === language;
    });
  };

  const toggleSelected = (
    e: React.ChangeEvent<HTMLInputElement>,
    language: string
  ) => {
    onLanguageChange(e, language);
  };

  return (
    <fieldset className="language-accordion">
      <legend>
        <DS.Label htmlFor="lanaguage-select">Languages</DS.Label>
      </legend>
      <DS.Accordion
        inputId="language-select"
        accordionLabel="Filter Languages"
        defaultOpen={true}
        modifiers={["fixed-height"]}
      >
        <div style={{ height: "300px" }}>
          <DS.List
            id="languages-list"
            className="languages-list"
            type={DS.ListTypes.Unordered}
            modifiers={["no-list-styling"]}
          >
            {languages.map((language) => {
              return (
                <li key={`check-${language.value}`}>
                  <DS.Checkbox
                    name={language.value}
                    checkboxId={`checkbox-${language.value}`}
                    labelOptions={{
                      id: `checkbox-label-${language.value}`,
                      labelContent: (
                        <>
                          {language.value}{" "}
                          {showCount ? `(${language.count})` : ""}
                        </>
                      ),
                    }}
                    checked={!!selectedLanguageFilter(language.value)}
                    onChange={(e) => toggleSelected(e, language.value)}
                    attributes={{
                      "aria-labelledby": `checkbox-label-${language.value}`,
                    }}
                  />
                </li>
              );
            })}
          </DS.List>
        </div>
      </DS.Accordion>
    </fieldset>
  );
};

export default LanguageAccordion;
