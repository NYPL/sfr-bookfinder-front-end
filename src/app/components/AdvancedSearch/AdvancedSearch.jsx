/* eslint-disable jsx-a11y/label-has-associated-control */
import React from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import Select from 'react-select';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { withRouter } from 'react-router';

import appConfig from '../../../../appConfig';
import * as searchActions from '../../actions/SearchActions';
import Breadcrumbs from '../Breadcrumbs/Breadcrumbs';
import { getQueryString } from '../../search/query';
import { initialSearchQuery, searchQueryPropTypes } from '../../stores/InitialState';
import TextInput from '../Form/TextInput';
import Checkbox from '../Form/Checkbox';
import { inputTerms, formatTypes, errorMessagesText } from '../../constants/labels';
import FilterYears from '../SearchResults/FilterYears';


const initialState = {
  error: false,
  errorMsg: '',
  languages: [],
  queries: {
    keyword: '',
    title: '',
    author: '',
    subject: '',
  },
  filters: {
    format: {
      epub: false,
      pdf: false,
      html: false,
    },
    language: [],
    years: {
      start: '',
      end: '',
    },
  },
};
// style for the languages dropdown
const customStyles = {
  control: base => ({
    ...base,
    minHeight: '2.5rem',
  }),
  menu: base => ({
    ...base,
    background: '#FFF',
    // override border radius to match the box
    borderRadius: 0,
    border: '1px solid black',
    width: '100%',
    // kill the gap
    marginTop: 0,
    marginRigth: '2rem',
  }),
  menuList: base => ({
    ...base,
    // kill the white space on first and last option
    padding: 0,
  }),
  option: (base, state) => ({
    ...base,
    background: state.isFocused ? 'black' : 'white',
    color: state.isFocused ? 'white' : 'black',
  }),
};

/**
 * Container class providing the Redux action creators
 * to its child components. State data is passed along
 * including params set in location.
 *
 * Accessibility Note: Creates the <main> element for all
 * search pages with the corresponding <h1>.
 */
class AdvancedSearch extends React.Component {
  constructor(props) {
    super(props);
    const { dispatch } = props;
    this.state = initialState;
    this.boundActions = bindActionCreators(searchActions, dispatch);
    this.onQueryChange = this.onQueryChange.bind(this);
    this.onLanguageChange = this.onLanguageChange.bind(this);
    this.parseStateToQuery = this.parseStateToQuery.bind(this);
    this.submitSearchRequest = this.submitSearchRequest.bind(this);
    this.onChangeYears = this.onChangeYears.bind(this);
    this.onFormatChange = this.onFormatChange.bind(this);
    this.clearForm = this.clearForm.bind(this);
  }

  componentDidMount() {
    this.loadLanguages();
  }

  componentWillReceiveProps(props) {
    this.parseQueryToState(props.searchQuery);
  }

  onQueryChange(event) {
    const target = event.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const name = target.name;
    this.setState((prevState) => {
      const queries = prevState.queries;
      queries[name] = value;
      return { queries, error: false, errorMsg: '' };
    });
  }

  onLanguageChange(event) {
    const value = event && !event.target && event.map(language => language.value);
    if (value) {
      this.setState((prevState) => {
        const filters = prevState.filters;
        filters.language = value;
        return { filters };
      });
    }
  }

  onFormatChange(event) {
    const target = event.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const name = target.name;
    this.setState((prevState) => {
      const filters = prevState.filters;
      filters.format[name] = value;
      return { filters };
    });
  }

  onChangeYears(yearsFilter) {
    this.setState((prevState) => {
      const filters = prevState.filters;
      filters.years = yearsFilter;
      return { filters };
    });
  }

  onErrorYears(errorObj) {
    this.setState({ error: errorObj.error, errorMsg: errorObj.errorMsg });
  }

  loadLanguages() {
    const appEnv = process.env.APP_ENV || 'production';
    const apiUrl = appConfig.api[appEnv];
    const { languagesPath } = appConfig.api;
    const langUrl = apiUrl + languagesPath;

    axios.get(langUrl).then((resp) => {
      this.setState({
        languages:
          resp
          && resp.data
          && resp.data.data
          && resp.data.data.languages.map(language => ({ value: language.language, label: language.language })),
      });
    });
  }

  parseStateToQuery() {
    const queries = Object.keys(this.state.queries)
      .map((q) => {
        const query = this.state.queries[q] && this.state.queries[q].replace(/^\s+/, '').replace(/\s+$/, '');
        return { field: q, query };
      })
      .filter(q => !!q.query);

    const filters = [];
    if (this.state.filters.language && this.state.filters.language.length > 0) {
      Object.keys(this.state.filters.language).forEach((language) => {
        filters.push({ field: 'language', value: this.state.filters.language[language] });
      });
    }
    if (this.state.filters.years.start || this.state.filters.years.end) {
      filters.push({
        field: 'years',
        value: {
          start: this.state.filters.years.start || '',
          end: this.state.filters.years.end || '',
        },
      });
    }
    if (Object.keys(this.state.filters.format).length > 0) {
      Object.keys(this.state.filters.format).forEach((format) => {
        if (this.state.filters.format[format]) {
          filters.push({
            field: 'format',
            value: format,
          });
        }
      });
    }

    if ((!queries || queries.length < 1) && (!filters || filters.length < 1)) {
      return null;
    }
    return Object.assign({}, { page: '0', per_page: '10', sort: [] }, { queries, filters });
  }

  parseQueryToState(searchQuery) {
    const state = { queries: {}, filters: {} };
    if (!searchQuery) {
      return;
    }
    if (searchQuery.showField && searchQuery.showQuery) {
      state.queries[searchQuery.showField] = searchQuery.showQuery;
    }

    if (searchQuery.queries) {
      searchQuery.queries.forEach((q) => {
        const allowedFields = inputTerms.map(term => term.values.map(value => value.key)).flat();
        if (allowedFields.indexOf(q.field) >= 0) {
          // If field is already set (by showQuery), use the showQuery value.
          if (!state.queries[q.field]) {
            state.queries[q.field] = q.query;
          }
        } else {
          state.queries[q.field] = '';
        }
      });
    }
    if (searchQuery.filters) {
      searchQuery.filters.forEach((q) => {
        if (q.field === 'format') {
          if (!state.filters[q.field]) {
            state.filters[q.field] = {};
          }
          state.filters[q.field][q.value] = true;
        } else if (q.field === 'language') {
          if (!state.filters[q.field]) {
            state.filters[q.field] = [];
          }
          state.filters[q.field].push(q.value);
        } else {
          state.filters[q.field] = q.value;
        }
      });
    }
    this.setState((prevState) => {
      const filters = { ...prevState.filters, ...state.filters };
      const queries = { ...prevState.queries, ...state.queries };
      return { filters, queries };
    });
  }

  submitSearchRequest(event) {
    event.preventDefault();
    event.stopPropagation();
    if (this.state.filters && this.state.filters.years
      && this.state.filters.years.start && this.state.filters.years.end
      && Number(this.state.filters.years.start) > Number(this.state.filters.years.end)) {
      this.setState({ error: true, errorMsg: errorMessagesText.invalidDate });
      return;
    }
    const fullQuery = this.parseStateToQuery();
    if (!fullQuery || !fullQuery.queries || fullQuery.queries.length < 1) {
      this.setState({ error: true, errorMsg: errorMessagesText.emptySearch });
      return;
    }
    const path = `/search?${getQueryString(fullQuery)}`;
    this.context.router.push(path);
  }

  clearForm() {
    this.setState(initialState);
  }

  render() {
    const { searchQuery, location } = this.props;
    const { router } = this.context;
    const languagesSelected = this.state.languages.filter(language => this.state.filters.language.indexOf(language.value) > -1);
    const getQueryValue = key => this.state.queries[key];
    const getFilterValue = (filter, key) => this.state.filters[filter] && this.state.filters[filter][key];

    return (
      <main
        id="mainContent"
        className="main-content grid-container padding-0"
      >
        <div className="grid-row">
          <div className="sfr-header-wrapper tablet:grid-col-12">
            <Breadcrumbs
              router={router}
              location={location}
            />
          </div>
          <div
            aria-label="Digital Research Books Beta"
            className="grid-col-12"
          >
            <div className="sfr-header-wrapper grid-col-10">
              <h1 className="nypl-heading">Digital Research Books Beta</h1>
              <div id="tagline">Search the world&apos;s research collections and more for digital books you can use right now</div>
            </div>
            <div className="grid-col-10 margin-bottom-2 margin-x-auto">
              <form
                className="usa-form grid-container width-full margin-x-0 padding-x-0"
                onSubmit={this.submitSearchRequest}
                onKeyPress={(event) => { if (event.keyCode === 13) { this.submitSearchRequest(); } }}
              >
                <fieldset className="usa-fieldset ">
                  <legend className="usa-legend usa-sr-only">Advanced Search</legend>
                  {inputTerms.map(line => (
                    <div
                      className="grid-row grid-gap"
                      key={line.key}
                    >
                      {line.values.map(term => (
                        <TextInput
                          className="tablet:grid-col-6"
                          ariaLabel={`Search for ${term.label}`}
                          labelClass="usa-label"
                          id={term.key}
                          type="text"
                          inputClass="usa-input"
                          name={term.key}
                          onChange={this.onQueryChange}
                          label={term.label}
                          key={term.key}
                          value={getQueryValue(term.key)}
                        />
                      ))}
                    </div>
                  ))}
                </fieldset>

                <fieldset className="usa-fieldset">
                  <legend className="usa-legend usa-sr-only">Language Search</legend>
                  <label
                    htmlFor="language"
                    className="usa-label width-full margin-bottom-1"
                  >
                    Language
                  </label>
                  <div className="grid-row nypl-select">
                    <Select
                      options={this.state.languages}
                      className="tablet:grid-col-6 position-relative padding-x-0 margin-x-0"
                      isMulti
                      labelClass="usa-label"
                      selectClass="usa-select multiple"
                      name="language"
                      id="filters-language"
                      label="Languages"
                      value={languagesSelected}
                      onChange={this.onLanguageChange}
                      theme={theme => ({
                        ...theme,
                        borderRadius: 0,
                        colors: {
                          ...theme.colores,
                          primary: 'black',
                        },
                      })}
                      styles={customStyles}
                      classNamePrefix="nypl-select"
                    />
                    <div className="tablet:grid-col-6" />
                  </div>
                </fieldset>
                <div className="grid-row margin-top-4 grid-gap">
                  <div className="tablet:grid-col-6">
                    <fieldset className="usa-fieldset grid-container width-full margin-x-0 padding-x-0 margin-bottom-2">
                      <legend className="usa-legend font-sans-lg sub-legend">Publication Year</legend>
                      <FilterYears
                        searchQuery={searchQuery}
                        onChange={e => this.onChangeYears(e)}
                        onError={e => this.onErrorYears(e)}
                        inputClassName="tablet:grid-col padding-right-0 padding-top-2"
                        className="grid-row grid-gap"
                      />
                    </fieldset>
                  </div>

                  <div className="tablet:grid-col-6">
                    <fieldset className="usa-fieldset grid-container width-full margin-x-0 padding-x-0">
                      <legend className="usa-legend font-sans-lg sub-legend margin-bottom-3">Format</legend>
                      {formatTypes.map(formatType => (
                        <Checkbox
                          className="usa-checkbox tablet:grid-col-12"
                          labelClass="usa-checkbox__label"
                          inputClass="usa-checkbox__input"
                          id={`filters-format-${formatType.value}`}
                          isSelected={getFilterValue('format', formatType.value)}
                          onChange={this.onFormatChange}
                          label={formatType.label}
                          name={formatType.value}
                          key={`facet-format-${formatType.value}`}
                        />
                      ))}
                    </fieldset>
                  </div>
                </div>
                <div className="grid-row grid-gap">
                  <div className="tablet:grid-col-6">
                    <div className="grid-row grid-gap">
                      <div className="tablet:grid-col-6">
                        <input
                          className="usa-button width-full margin-top-1"
                          type="submit"
                          value="Search"
                          readOnly
                          onClick={this.submitSearchRequest}
                        />
                      </div>
                      <div className="tablet:grid-col-6">
                        <input
                          className="usa-button usa-button--outline width-full margin-top-1"
                          type="clear"
                          value="Clear"
                          onClick={this.clearForm}
                          readOnly
                        />
                      </div>
                    </div>
                    {this.state.error && (
                      <div
                        className="usa-alert usa-alert--error"
                        role="alert"
                      >
                        <div className="usa-alert__body">
                          <h3 className="usa-alert__heading">Error</h3>
                          <p className="usa-alert__text">{this.state.errorMsg}</p>
                        </div>
                      </div>
                    )}
                  </div>
                </div>
              </form>
            </div>
          </div>
        </div>
      </main>
    );
  }
}

AdvancedSearch.propTypes = {
  location: PropTypes.objectOf(PropTypes.any),
  searchQuery: searchQueryPropTypes,
  dispatch: PropTypes.func,
};

AdvancedSearch.defaultProps = {
  location: {},
  searchQuery: initialSearchQuery,
  dispatch: () => { },
};

AdvancedSearch.contextTypes = {
  router: PropTypes.objectOf(PropTypes.any),
  history: PropTypes.objectOf(PropTypes.any),
};

const mapStateToProps = state => state;

export default connect(
  mapStateToProps,
  null,
)(withRouter(AdvancedSearch));
