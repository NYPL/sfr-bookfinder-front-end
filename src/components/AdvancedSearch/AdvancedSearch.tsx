import React, { useState } from "react";
import { useRouter } from "next/router";

import {
  findFiltersForField,
  findFiltersExceptField,
  findQueryForField,
} from "~/src/util/SearchQueryUtils";
import { initialSearchQuery } from "~/src/constants/InitialState";
import {
  inputTerms,
  errorMessagesText,
  breadcrumbTitles,
} from "~/src/constants/labels";
import FilterYears from "~/src/components/FilterYears/FilterYears";
import { SearchQuery, SearchQueryDefaults } from "~/src/types/SearchQuery";

import * as DS from "@nypl/design-system-react-components";
import LanguageAccordion from "../LanguageAccordion/LanguageAccordion";
import FilterBookFormat from "../FilterBookFormat/FilterBookFormat";
import { FacetItem } from "~/src/types/DataModel";
import { toLocationQuery } from "~/src/util/SearchUtils";
import { toApiQuery } from "~/src/util/apiConversion";

const AdvancedSearch: React.FC<{
  searchQuery: SearchQuery;
  languages: FacetItem[];
}> = (props) => {
  const { languages } = props;
  const router = useRouter();
  const [searchQuery, setSearchQuery] = useState({
    ...SearchQueryDefaults,
    ...props.searchQuery,
  });
  const [emptySearchError, setEmptySearchError] = useState("");
  const [dateRangeError, setDateRangeError] = useState("");

  const submit = (e) => {
    e.preventDefault();
    if (!searchQuery.queries || searchQuery.queries.length < 1) {
      setEmptySearchError(errorMessagesText.emptySearch);
      return;
    } else {
      setEmptySearchError("");
    }

    if (
      searchQuery.filterYears.start &&
      searchQuery.filterYears.end &&
      searchQuery.filterYears.end < searchQuery.filterYears.start
    ) {
      setDateRangeError(errorMessagesText.invalidDate);
      return;
    } else {
      setDateRangeError("");
    }

    router.push({
      pathname: "/search",
      query: toLocationQuery(toApiQuery(searchQuery)),
    });
  };

  const clearSearch = () => {
    setSearchQuery(initialSearchQuery);
  };

  const onQueryChange = (e, queryKey) => {
    const newQuery = {
      field: queryKey,
      query: e.target.value,
    };

    const allQueries = searchQuery.queries.filter((query) => {
      return query.field !== queryKey;
    });

    allQueries.push(newQuery);
    setSearchQuery({
      ...searchQuery,
      queries: allQueries,
    });
  };

  const onLanguageChange = (e, language) => {
    const languageFilters = findFiltersForField(
      searchQuery.filters,
      "language"
    );

    setSearchQuery({
      ...searchQuery,
      filters: [
        ...findFiltersExceptField(searchQuery.filters, "language"),
        ...(e.target.checked
          ? [...languageFilters, { field: "language", value: language }]
          : languageFilters.filter((filter) => {
              return filter.value !== language;
            })),
      ],
    });
  };

  const onBookFormatChange = (e, format) => {
    const formatFilters = findFiltersForField(searchQuery.filters, "format");

    setSearchQuery({
      ...searchQuery,
      filters: [
        ...findFiltersExceptField(searchQuery.filters, "format"),
        ...(e.target.checked
          ? [...formatFilters, { field: "format", value: format }]
          : formatFilters.filter((filter) => {
              return filter.value !== format;
            })),
      ],
    });
  };

  const onDateChange = (e, isStart: boolean) => {
    console.log("searchQuery", searchQuery);
    setSearchQuery({
      ...searchQuery,
      filterYears: {
        start: isStart ? e.target.value : searchQuery.filterYears.start,
        end: isStart ? searchQuery.filterYears.end : e.target.value,
      },
    });
  };

  return (
    <main id="mainContent" className="main">
      <div className="content-header">
        <DS.Breadcrumbs
          breadcrumbs={[{ url: "/", text: breadcrumbTitles.home }]}
        />{" "}
      </div>

      <div aria-label="Digital Research Books Beta" className="grid-col-12">
        <div className="sfr-header-wrapper grid-col-10">
          <h1 className="nypl-heading">Digital Research Books Beta</h1>

          <div id="tagline">
            Search the world&apos;s research collections and more for digital
            books you can use right now
          </div>
        </div>

        <form
          className="usa-form grid-container width-full margin-x-0 padding-x-0"
          onSubmit={(e) => {
            submit(e);
          }}
        >
          {emptySearchError && (
            <DS.HelperErrorText isError={true}>
              {emptySearchError}
            </DS.HelperErrorText>
          )}
          {dateRangeError && (
            <DS.HelperErrorText isError={true}>
              {dateRangeError}
            </DS.HelperErrorText>
          )}
          <fieldset>
            <legend>Advanced Search</legend>

            {inputTerms.map((field: { key: string; label: string }) => {
              return (
                <div key={`field-${field.key}`}>
                  <DS.Label
                    htmlFor={`${field.key}-input`}
                    id={`${field.key}-label`}
                  >
                    {field.label}
                  </DS.Label>
                  <DS.Input
                    id={`${field.key}-input`}
                    ariaLabel={`Input for ${field.label}`}
                    value={
                      findQueryForField(searchQuery.queries, field.key)
                        ? findQueryForField(searchQuery.queries, field.key)
                            .query
                        : ""
                    }
                    onChange={(e) => onQueryChange(e, field.key)}
                  />
                </div>
              );
            })}
          </fieldset>
          {languages.length > 0 && (
            <LanguageAccordion
              languages={languages}
              showCount={false}
              selectedLanguages={findFiltersForField(
                searchQuery.filters,
                "language"
              )}
              onLanguageChange={(e, language) => onLanguageChange(e, language)}
            />
          )}
          <FilterYears
            dateFilters={searchQuery.filterYears}
            onDateChange={(e, isStart) => {
              onDateChange(e, isStart);
            }}
          />
          <FilterBookFormat
            selectedFormats={findFiltersForField(searchQuery.filters, "format")}
            onFormatChange={(e, format) => {
              onBookFormatChange(e, format);
            }}
          />
          <DS.Button
            id="search-button"
            buttonType={DS.ButtonTypes.Primary}
            type="submit"
          >
            Search
          </DS.Button>
          <DS.Button
            id="clear-button"
            buttonType={DS.ButtonTypes.Secondary}
            type="reset"
            onClick={() => clearSearch()}
          >
            Clear
          </DS.Button>
        </form>
      </div>
    </main>
  );
};

export default AdvancedSearch;
