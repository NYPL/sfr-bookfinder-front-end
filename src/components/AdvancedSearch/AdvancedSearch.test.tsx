import React from "react";
import AdvancedSearch from "./AdvancedSearch";
import { initialSearchQuery } from "~/src/constants/InitialState";
import { FacetItem } from "~/src/types/DataModel";
import { SearchQuery } from "~/src/types/SearchQuery";
import { shallow, mount } from "enzyme";
import LanguageAccordion from "~/src/components/LanguageAccordion/LanguageAccordion";
import {
  MockNextRouterContextProvider,
  mockPush,
} from "~/src/__tests__/testUtils/MockNextRouter";
import { HelperErrorText } from "@nypl/design-system-react-components";

const defaultLanguages: FacetItem[] = [
  { value: "english", count: 25 },
  { value: "french", count: 30 },
];

const complicatedSearchQuery: SearchQuery = {
  perPage: 10,
  page: 0,
  filters: [{ value: "english", field: "language" }],
  filterYears: { start: 1990, end: 1999 },
  sort: { field: "relevance", dir: "DESC" },
  queries: [{ field: "keyword", query: "cat" }],
  showAll: false,
};

describe("renders advanced search correctly", () => {
  test("Renders correctly in when passed well-formed query", () => {
    const advancedSearch = shallow(
      <AdvancedSearch
        searchQuery={initialSearchQuery}
        languages={defaultLanguages}
      />
    );
    expect(advancedSearch).toMatchSnapshot();
  });

  test("Hides languages when no languages are passed", () => {
    const wrapper = shallow(
      <AdvancedSearch searchQuery={initialSearchQuery} languages={[]} />
    );
    expect(wrapper.exists(LanguageAccordion)).toEqual(false);
  });

  test("Prepopulates form input based on search query parameters", () => {
    const wrapper = mount(
      <AdvancedSearch
        searchQuery={complicatedSearchQuery}
        languages={defaultLanguages}
      />
    );

    expect(wrapper.find("#checkbox-english").props().checked).toEqual(true);
    expect(wrapper.find("input#date-filter-from").props().value).toEqual(1990);
    expect(wrapper.find("input#date-filter-to").props().value).toEqual(1999);
    expect(wrapper.find("input#keyword-input").props().value).toEqual("cat");
  });
});

describe("Advanced Search submit", () => {
  test("Submits well formed query", () => {
    const wrapper = mount(
      <MockNextRouterContextProvider>
        <AdvancedSearch
          searchQuery={complicatedSearchQuery}
          languages={defaultLanguages}
        />
      </MockNextRouterContextProvider>
    );
    wrapper.find("form").simulate("submit");

    const expectedQuery = {
      filters: `[{"field":"language","value":"english"},{"field":"years","value":{"start":1990,"end":1999}},{"field":"show_all","value":false}]`,
      queries: `[{"field":"keyword","query":"cat"}]`,
      per_page: "10",
      sort: `[]`,
    };
    expect(mockPush).toHaveBeenCalledTimes(1);
    expect(mockPush).toHaveBeenCalledWith({
      pathname: "/search",
      query: expectedQuery,
    });
  });

  test("shows error on empty query", () => {
    const wrapper = mount(
      <MockNextRouterContextProvider>
        <AdvancedSearch
          searchQuery={initialSearchQuery}
          languages={defaultLanguages}
        />
      </MockNextRouterContextProvider>
    );
    wrapper.find("form").simulate("submit");
    expect(mockPush).toHaveBeenCalledTimes(0);
    expect(wrapper.exists(HelperErrorText)).toEqual(true);
    expect(wrapper.find(HelperErrorText)).toMatchSnapshot();
  });

  test("show error on invalid year", () => {
    const invalidYearSearch = Object.assign({}, initialSearchQuery, {
      queries: [{ field: "keyword", query: "cat" }],
      filterYears: { start: 1990, end: 1880 },
    });
    const wrapper = mount(
      <MockNextRouterContextProvider>
        <AdvancedSearch
          searchQuery={invalidYearSearch}
          languages={defaultLanguages}
        />
      </MockNextRouterContextProvider>
    );
    wrapper.find("form").simulate("submit");
    expect(mockPush).toHaveBeenCalledTimes(0);
    expect(wrapper.exists(HelperErrorText)).toEqual(true);
    expect(wrapper.find(HelperErrorText)).toMatchSnapshot();
  });
});

describe("Advanced Search clear", () => {
  test("Clears search", () => {
    const wrapper = mount(
      <MockNextRouterContextProvider>
        <AdvancedSearch
          searchQuery={complicatedSearchQuery}
          languages={defaultLanguages}
        />
      </MockNextRouterContextProvider>
    );
    expect(wrapper.find("#checkbox-english").props().checked).toEqual(true);

    wrapper.find(`#clear-button`).hostNodes().simulate("click");
    expect(wrapper.find("#checkbox-english").props().checked).toEqual(false);
  });
});
