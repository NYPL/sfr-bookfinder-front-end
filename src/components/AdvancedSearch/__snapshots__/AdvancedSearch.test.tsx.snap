// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Advanced Search submit show error on invalid year 1`] = `
Array [
  <HelperErrorText
    isError={true}
  >
    <div
      aria-atomic={true}
      aria-live="polite"
      className="helper-text helper-text--error"
    >
      Start date must be before End date
    </div>
  </HelperErrorText>,
  <HelperErrorText
    isError={false}
  >
    <div
      aria-atomic={true}
      aria-live="off"
      className="helper-text"
    >
       EX. 1901 
    </div>
  </HelperErrorText>,
  <HelperErrorText
    isError={false}
  >
    <div
      aria-atomic={true}
      aria-live="off"
      className="helper-text"
    >
       EX. 2000 
    </div>
  </HelperErrorText>,
]
`;

exports[`Advanced Search submit shows error on empty query 1`] = `
Array [
  <HelperErrorText
    isError={true}
  >
    <div
      aria-atomic={true}
      aria-live="polite"
      className="helper-text helper-text--error"
    >
      Please enter a search term
    </div>
  </HelperErrorText>,
  <HelperErrorText
    isError={false}
  >
    <div
      aria-atomic={true}
      aria-live="off"
      className="helper-text"
    >
       EX. 1901 
    </div>
  </HelperErrorText>,
  <HelperErrorText
    isError={false}
  >
    <div
      aria-atomic={true}
      aria-live="off"
      className="helper-text"
    >
       EX. 2000 
    </div>
  </HelperErrorText>,
]
`;

exports[`renders advanced search correctly Renders correctly in when passed well-formed query 1`] = `
<main
  className="main"
  id="mainContent"
>
  <div
    className="content-header"
  >
    <Breadcrumbs
      breadcrumbs={
        Array [
          Object {
            "text": "Digital Research Books Beta",
            "url": "/",
          },
        ]
      }
    />
     
  </div>
  <div
    aria-label="Digital Research Books Beta"
    className="grid-col-12"
  >
    <div
      className="sfr-header-wrapper grid-col-10"
    >
      <h1
        className="nypl-heading"
      >
        Digital Research Books Beta
      </h1>
      <div
        id="tagline"
      >
        Search the world's research collections and more for digital books you can use right now
      </div>
    </div>
    <form
      className="usa-form grid-container width-full margin-x-0 padding-x-0"
      onSubmit={[Function]}
    >
      <fieldset>
        <legend>
          Advanced Search
        </legend>
        <div
          key="field-keyword"
        >
          <Label
            htmlFor="keyword-input"
            id="keyword-label"
          >
            Keyword
          </Label>
          <Input
            ariaLabel="Input for Keyword"
            id="keyword-input"
            onChange={[Function]}
            value=""
          />
        </div>
        <div
          key="field-author"
        >
          <Label
            htmlFor="author-input"
            id="author-label"
          >
            Author
          </Label>
          <Input
            ariaLabel="Input for Author"
            id="author-input"
            onChange={[Function]}
            value=""
          />
        </div>
        <div
          key="field-title"
        >
          <Label
            htmlFor="title-input"
            id="title-label"
          >
            Title
          </Label>
          <Input
            ariaLabel="Input for Title"
            id="title-input"
            onChange={[Function]}
            value=""
          />
        </div>
        <div
          key="field-subject"
        >
          <Label
            htmlFor="subject-input"
            id="subject-label"
          >
            Subject
          </Label>
          <Input
            ariaLabel="Input for Subject"
            id="subject-input"
            onChange={[Function]}
            value=""
          />
        </div>
      </fieldset>
      <LanguageAccordion
        languages={
          Array [
            Object {
              "count": 25,
              "value": "english",
            },
            Object {
              "count": 30,
              "value": "french",
            },
          ]
        }
        onLanguageChange={[Function]}
        selectedLanguages={Array []}
        showCount={false}
      />
      <FilterYears
        dateFilters={
          Object {
            "end": null,
            "start": null,
          }
        }
        onDateChange={[Function]}
      />
      <FilterBookFormat
        onFormatChange={[Function]}
        selectedFormats={Array []}
      />
      <Button
        buttonType="filled"
        id="search-button"
        mouseDown={false}
        type="submit"
      >
        Search
      </Button>
      <Button
        buttonType="outline"
        id="clear-button"
        mouseDown={false}
        onClick={[Function]}
        type="reset"
      >
        Clear
      </Button>
    </form>
  </div>
</main>
`;
